File: backend/app/controllers/history_controller.py
from datetime import datetime, timezone

from bson import ObjectId
from fastapi import HTTPException, status

from backend.app.db.mongo import get_database
from backend.app.models.history_model import HistoryCreateRequest


def _to_object_id(id_value: str, entity_name: str) -> ObjectId:
    try:
        return ObjectId(id_value)
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid {entity_name} id format.",
        ) from exc


def _serialize_history(document: dict) -> dict:
    return {
        "id": str(document["_id"]),
        "user_id": str(document["user_id"]),
        "file_name": document["file_name"],
        "prediction": document["prediction"],
        "confidence": float(document["confidence"]),
        "real_score": float(document["real_score"]),
        "fake_score": float(document["fake_score"]),
        "timestamp": document["timestamp"].isoformat(),
    }


async def create_history_entry(payload: HistoryCreateRequest) -> dict:
    db = get_database()
    users = db.users
    history = db.history

    user_oid = _to_object_id(payload.user_id, "user")
    user = await users.find_one({"_id": user_oid})
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found for history entry.",
        )

    timestamp = (
        datetime.fromisoformat(payload.timestamp.replace("Z", "+00:00"))
        if payload.timestamp
        else datetime.now(timezone.utc)
    )

    document = {
        "user_id": user_oid,
        "file_name": payload.file_name,
        "prediction": payload.prediction,
        "confidence": payload.confidence,
        "real_score": payload.real_score,
        "fake_score": payload.fake_score,
        "timestamp": timestamp,
        "created_at": datetime.now(timezone.utc),
    }

    result = await history.insert_one(document)
    created = await history.find_one({"_id": result.inserted_id})
    return _serialize_history(created)


async def get_history_for_user(user_id: str) -> list[dict]:
    db = get_database()
    history = db.history

    user_oid = _to_object_id(user_id, "user")

    cursor = history.find({"user_id": user_oid}).sort("timestamp", -1)
    documents = await cursor.to_list(length=2000)
    return [_serialize_history(doc) for doc in documents]


async def delete_history_entry(entry_id: str) -> dict:
    db = get_database()
    history = db.history

    entry_oid = _to_object_id(entry_id, "history entry")
    result = await history.delete_one({"_id": entry_oid})

    if result.deleted_count == 0:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="History entry not found.",
        )

    return {"deleted": True, "entry_id": entry_id}


async def clear_history_for_user(user_id: str) -> dict:
    db = get_database()
    history = db.history

    user_oid = _to_object_id(user_id, "user")
    result = await history.delete_many({"user_id": user_oid})

    return {"deleted": int(result.deleted_count), "user_id": user_id}

File: Web_Backend/controllers/historyController.js
const mongoose = require("mongoose");

const History = require("../models/History");
const User = require("../models/User");

const serializeHistory = (historyDoc) => ({
  id: historyDoc._id.toString(),
  user_id: historyDoc.user_id.toString(),
  file_name: historyDoc.file_name,
  prediction: historyDoc.prediction,
  confidence: Number(historyDoc.confidence),
  real_score: Number(historyDoc.real_score),
  fake_score: Number(historyDoc.fake_score),
  timestamp: historyDoc.timestamp.toISOString(),
});

const isValidScore = (value) =>
  typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1;

const createHistoryEntry = async (req, res) => {
  try {
    const {
      user_id,
      file_name,
      prediction,
      confidence,
      real_score,
      fake_score,
      timestamp,
    } = req.body;

    if (!user_id || !file_name || !prediction) {
      return res.status(400).json({ detail: "user_id, file_name and prediction are required." });
    }

    if (
      !isValidScore(Number(confidence)) ||
      !isValidScore(Number(real_score)) ||
      !isValidScore(Number(fake_score))
    ) {
      return res.status(400).json({ detail: "Score values must be between 0 and 1." });
    }

    const user = await User.findById(user_id);
    if (!user) {
      return res.status(404).json({ detail: "User not found for history entry." });
    }

    const entry = await History.create({
      user_id: user._id,
      file_name: String(file_name).trim(),
      prediction: String(prediction),
      confidence: Number(confidence),
      real_score: Number(real_score),
      fake_score: Number(fake_score),
      timestamp: timestamp ? new Date(timestamp) : new Date(),
    });

    return res.status(201).json(serializeHistory(entry));
  } catch (error) {
    return res.status(400).json({ detail: "Failed to create history entry." });
  }
};

const getHistoryByUser = async (req, res) => {
  try {
    const { userId } = req.params;

    if (!mongoose.isValidObjectId(userId)) {
      return res.status(400).json({ detail: "Invalid user id." });
    }

    const entries = await History.find({ user_id: userId }).sort({ timestamp: -1 });
    return res.status(200).json(entries.map(serializeHistory));
  } catch (error) {
    return res.status(500).json({ detail: "Failed to fetch history." });
  }
};

const deleteHistoryEntry = async (req, res) => {
  try {
    const { entryId } = req.params;

    if (!mongoose.isValidObjectId(entryId)) {
      return res.status(400).json({ detail: "Invalid history id." });
    }

    const deleted = await History.findByIdAndDelete(entryId);
    if (!deleted) {
      return res.status(404).json({ detail: "History entry not found." });
    }

    return res.status(200).json({ deleted: true, entry_id: entryId });
  } catch (error) {
    return res.status(500).json({ detail: "Failed to delete history entry." });
  }
};

const clearHistoryByUser = async (req, res) => {
  try {
    const { userId } = req.params;

    if (!mongoose.isValidObjectId(userId)) {
      return res.status(400).json({ detail: "Invalid user id." });
    }

    const result = await History.deleteMany({ user_id: userId });
    return res.status(200).json({ deleted: result.deletedCount, user_id: userId });
  } catch (error) {
    return res.status(500).json({ detail: "Failed to clear history." });
  }
};

module.exports = {
  createHistoryEntry,
  getHistoryByUser,
  deleteHistoryEntry,
  clearHistoryByUser,
};

